### ISSUE TITLE
transfer the target image's points to source image?

### ISSUE THREAD

**Issue by willer94** [SunghwanHong/CATs#5]
In your training  and testing code, the target image's points are transfered  into source image, which is different to the setting in DHPF and CHM.
Is there anything wrong?

**Comment by SunghwanHong** [comment]
Hi
Could you please specify which line in which file you found strange?

**Comment by willer94** [comment]
Speciﬁcally， in the evaluation.py, line 32:

```python

    @classmethod
    def eval_kps_transfer(cls, prd_kps, batch):
        r"""Compute percentage of correct key-points (PCK) based on prediction"""

        easy_match = {'src': [], 'trg': [], 'dist': []}
        hard_match = {'src': [], 'trg': []}

        pck = []
        for idx, (pk, tk) in enumerate(zip(prd_kps, batch['src_kps'])):
            thres = batch['pckthres'][idx]
            npt = batch['n_pts'][idx]
            _, correct_ids, _ = cls.classify_prd(pk[:, :npt], tk[:, :npt], thres)

            pck.append((len(correct_ids) / npt.item()) * 100)

        eval_result = {'pck': pck}

        return eval_result
```

In the code of dhpf, the [evaluation](https://github.com/juhongm999/dhpf/blob/31b8c9a9ccf71875b63b7e4c623cdb605c769463/common/evaluation.py#L20) is 
```python
    @classmethod
    def eval_kps_transfer(cls, prd_kps, batch):
        r"""Compute percentage of correct key-points (PCK) based on prediction"""

        easy_match = {'src': [], 'trg': [], 'dist': []}
        hard_match = {'src': [], 'trg': []}

        pck = []
        for idx, (pk, tk) in enumerate(zip(prd_kps, batch['trg_kps'])):
            thres = batch['pckthres'][idx]
            npt = batch['n_pts'][idx]
            correct_dist, correct_ids, incorrect_ids = cls.classify_prd(pk[:, :npt], tk[:, :npt], thres)

            # Collect easy and hard match feature index & store pck to buffer
            easy_match['dist'].append(correct_dist)
            easy_match['src'].append(batch['src_kpidx'][idx][:npt][correct_ids])
            easy_match['trg'].append(batch['trg_kpidx'][idx][:npt][correct_ids])
            hard_match['src'].append(batch['src_kpidx'][idx][:npt][incorrect_ids])
            hard_match['trg'].append(batch['trg_kpidx'][idx][:npt][incorrect_ids])
            pck.append((len(correct_ids) / npt.item()) * 100)

        eval_result = {'easy_match': easy_match,
                       'hard_match': hard_match,
                       'pck': pck}

        return eval_result
```

Also in CHM, the [evaluation](https://github.com/juhongm999/chm/blob/d17d819a2bd1b85a3ed758ad591825cbc7380db1/common/evaluation.py#L13) is 
```python
    @classmethod
    def evaluate(cls, prd_kps, batch):
        r""" Compute percentage of correct key-points (PCK) with multiple alpha {0.05, 0.1, 0.15 }"""

        pcks = []
        for idx, (pk, tk) in enumerate(zip(prd_kps, batch['trg_kps'])):
            pckthres = batch['pckthres'][idx].cuda()
            npt = batch['n_pts'][idx]
            prd_kps = pk[:, :npt].cuda()
            trg_kps = tk[:, :npt].cuda()

            l2dist = (prd_kps - trg_kps).pow(2).sum(dim=0).pow(0.5).unsqueeze(0).repeat(len(cls.alpha), 1)
            thres = pckthres.expand_as(l2dist).float() * cls.alpha
            pck = torch.le(l2dist, thres).sum(dim=1) / float(npt)
            if len(pck) == 1: pck = pck[0]
            pcks.append(pck)

        eval_result = {'pck': pcks}

        return eval_result
```
In CHM, the [evaluation](https://github.com/csyanbin/SCOT/blob/a2c7b96e777be2046742fe4441ad5f54608ef722/model/evaluation.py#L38) code is
```python 
    def evaluate(self, prd_kps, data):
        r"""Compute desired evaluation metric"""
        return self.eval_funct(prd_kps, data)

    def eval_pck(self, prd_kps, data):
        r"""Compute percentage of correct key-points (PCK) based on prediction"""
        pckthres = data['pckthres'][0] * data['trg_intratio']
        ncorrt = correct_kps(data['trg_kps'].cuda(), prd_kps, pckthres, data['alpha'])
        pair_pck = int(ncorrt) / int(data['trg_kps'].size(1))

        self.eval_buf['pck'].append(pair_pck)

        if self.eval_buf['cls_pck'].get(data['pair_class'][0]) is None:
            self.eval_buf['cls_pck'][data['pair_class'][0]] = []
        self.eval_buf['cls_pck'][data['pair_class'][0]].append(pair_pck)

        return pair_pck

def correct_kps(trg_kps, prd_kps, pckthres, alpha=0.1):
    r"""Compute the number of correctly transferred key-points"""
    l2dist = torch.pow(torch.sum(torch.pow(trg_kps - prd_kps, 2), 0), 0.5)
    thres = pckthres.expand_as(l2dist).float()
    correct_pts = torch.le(l2dist, thres * alpha)

    return torch.sum(correct_pts)

```
You can also check their testing code. 

It seems that the common practice is to warp the keypoints in a source image into a target image, and calculate the pck scores using the warpped location and the ground-truth in the target image. 
However, the keypoints in a target image are transferred into the source image in your evaluation code, can you please check if this is consistent with the common practice?

Many thanks!

**Comment by SunghwanHong** [comment]
for i, mini_batch in pbar:
            flow_gt = mini_batch['flow'].to(device)
            pred_flow = net(mini_batch['trg_img'].to(device),
                            mini_batch['src_img'].to(device))

            estimated_kps = flow2kps(mini_batch['trg_kps'].to(device), pred_flow, mini_batch['n_pts'].to(device))

            eval_result = Evaluator.eval_kps_transfer(estimated_kps.cpu(), mini_batch)



def test(model, dataloader):
    average_meter = AverageMeter(dataloader.dataset.benchmark)

    model.eval()
    for idx, batch in enumerate(dataloader):

        # 1. CHMNet forward pass
        corr_matrix = model(batch['src_img'].cuda(), batch['trg_img'].cuda())

        # 2. Transfer key-points
        prd_kps = Geometry.transfer_kps(corr_matrix, batch['src_kps'].cuda(), batch['n_pts'].cuda(), normalized=False)

        # 3. Evaluate predictions
        eval_result = Evaluator.evaluate(Geometry.unnormalize_kps(prd_kps), batch)
        average_meter.update(eval_result)
        average_meter.write_test_process(idx, len(dataloader))

    return average_meter.get_test_result()

**Comment by willer94** [comment]
Hi, thanks for your response.

However, I am still confused here. 

If I understand correctly, in your code,  the `estimated_kps`  is the projected  keypoints from a target images to a source 
image, using estimated flow.  Then the score is calculated in the function `Evaluator.eval_kps_transfer`, where the projected points are compared with the ground-truth in the **_source_** image. 

However, in CHM, the function `Geometry.transfer_kps` projects points in a source image into a target image, and the function `Evaluator.evaluate` calculates the distance between the projection and the ground-truth in the **_target_** image.

Please correct me if I am wrong.

Thanks!😊
